lib/transactions/sendAndConfirm.ts

```diff
@@ -0,0 +1,136 @@
+ * @param {TransactionLifecycleEventCallback} [params.onTransactionEvent] - Optional callback to handle transaction lifecycle events.
+ *
+ * @returns {Promise<string>} - A promise that resolves with the transaction ID if the transaction is confirmed successfully, or rejects with an error if the transaction fails or timeouts.
+ *
+ * @throws {ConfirmationTimeoutError} - Throws this error if the confirmation process times out.
+ * @throws {InstructionError} - Throws this error if there is an issue with the transaction's instructions.
+ * @throws {TransactionError} - Throws this error for general transaction failures.
+ */
+
+export const sendAndConfirmTransaction = async ({
+  signedTransaction,
+  connection,
+  config = {
+    type: 'static',
+    timeoutMs: DEFAULT_CONFIRMATION_TIMEOUT,
+  },
+  onTransactionEvent,
+}: {
+  signedTransaction: Transaction | VersionedTransaction;
+  connection: Connection;
+  config?: SendTransactionConfig &
+    (
+      | StaticTimeoutConfig
+      | TransactionExpirationTimeoutConfig
+      | NoTimeoutConfig
+    );
+  onTransactionEvent?: TransactionLifecycleEventCallback;
+}): Promise<string> => {
+  const startTime = getUnixTs();
+  const controller = new AbortController();
+
+  const {
+    continuouslySendTransactions,
+    pollingSendTransactionTimeoutMs,
+    sendOptions,
+    ...confirmationConfig
+  } = config;
+
+  const transactionId = await sendSignedTransaction({
+    signedTransaction,
+    connection,
+    config: {
+      controller,
+      continuouslySendTransactions,
+      pollingSendTransactionTimeoutMs,
+      sendOptions,
+    },
+    onTransactionEvent,
+  });
+
+  if (confirmationConfig.type === 'none') {
+    controller.abort();
+    log(
+      'Caller requested no confirmation, skipping all confirmation and returning after initial transaction sent to cluster'
+    );
+
+    return transactionId;
+  }
+
+  try {
+    await awaitTransactionSignatureConfirmation({
+      connection,
+      transactionId,
+      config: confirmationConfig,
+      onTransactionEvent,
+      controller,
+      transactionCommitment:
+        confirmationConfig?.type === 'expiration'
+          ? confirmationConfig?.transactionCommitment
+          : undefined,
+    });
+
+    log(
+      'Finished transaction status confirmation: ',
+      transactionId,
+      getUnixTs() - startTime
+    );
+  } catch (error: any) {
+    if (error.timeout) {
+      throw new ConfirmationTimeoutError({
+        transactionId,
+        message: 'Timed out awaiting confirmation on transaction',
+        config: ConfirmationTimeoutError.formatConfig(confirmationConfig),
+      });
+    } else if (error.err) {
+      throw new InstructionError({
+        transactionId,
+        error: error.err,
+        transaction: signedTransaction,
+      });
+    }
+
+    // note: if we want to try and get more information about a transaction failure, we could make a simulate transaction request here
```

Comment:
Does simulation afford us an estimate on compute? I want to get better with estimating that. Just a question.

Reply:
It looks like the simulateTransaction response _should_ contain units consumed (per [RPC docs](https://solana.com/docs/rpc/http/simulatetransaction)). I also manually checked on a locally running instance of the solana-test-validator (v1.16.20) and was able to see the unitsConsumed field in the output:

```bash
Transaction 5wVxmmGRqaeYDDbp1EeLzZR6HZ5ZcCbnmPWxxWBFDeARDYRinKa42ArkUeS97F9Tycq1vRxwP1Gg8pgQQRhb7GiH simulation result:  {
  accounts: null,
  err: { InstructionError: [ 3, 'IllegalOwner' ] },
  logs: [
    'Program 11111111111111111111111111111111 invoke [1]',
    'Program 11111111111111111111111111111111 success',
    'Program 11111111111111111111111111111111 invoke [1]',
    'Program 11111111111111111111111111111111 success',
    'Program 11111111111111111111111111111111 invoke [1]',
    'Program 11111111111111111111111111111111 success',
    'Program opnb2LAfJYbRMAHHvqjCwQxanZn7ReEHp1k81EohpZb invoke [1]',
    'Program log: Instruction: CreateMarket',
    'Program 11111111111111111111111111111111 invoke [2]',
    'Program 11111111111111111111111111111111 success',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL invoke [2]',
    'Program log: Create',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [3]',
    'Program log: Instruction: GetAccountDataSize',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 1622 of 971299 compute units',
    'Program return: TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA pQAAAAAAAAA=',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success',
    'Program 11111111111111111111111111111111 invoke [3]',
    'Program 11111111111111111111111111111111 success',
    'Program log: Initialize the associated token account',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [3]',
    'Program log: Instruction: InitializeImmutableOwner',
    'Program log: Please upgrade to SPL Token 2022 for immutable owner support',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 1405 of 964659 compute units',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [3]',
    'Program log: Instruction: InitializeAccount3',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 4241 of 960775 compute units',
    'Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL consumed 20544 of 976774 compute units',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL success',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL invoke [2]',
    'Program log: Create',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL consumed 3558 of 947720 compute units',
    'Program ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL failed: Provided owner is not allowed',
    'Program opnb2LAfJYbRMAHHvqjCwQxanZn7ReEHp1k81EohpZb consumed 55388 of 999550 compute units',
    'Program opnb2LAfJYbRMAHHvqjCwQxanZn7ReEHp1k81EohpZb failed: Provided owner is not allowed'
  ],
  returnData: null,
  unitsConsumed: 450
}
```

I already have a working function to plug in. I didn't push in the first round since it wasn't directly called, but I can add it and see what you think. So, it should definitely be possible to detect (+ log) instances in where transactions fail due to improper compute unit limits.

Reply:
Ok with the newest commit, if a transaction fails, we will attempt to simulate it. If the simulation is successful, we emit an event with the simulation result that contains the `unitsConsumed`. We also have the opportunity to try and parse logs if they exist.

Reply:
I now realize you might have meant "can we simulate and grab units consumed for every transaction", so apologies if I misunderstood. But, it looks we can incorporate if we want.